Implement xenstore permissions

LKomaryanskiy opened the pull request:
This PR mainly introduces of xenstore permissions feature. The feature is needed to allow or deny access between domains to process some operations with xenstore entries. In additionally to permissions support, also implemented the following:
1) Removing of entries that were created for domain or by domain in case of destruction;
2) Helper functions to create entries and set permission in one action;
3) Moved `initialize_xenstore` function to appropriate place;
4) Minor fix: /tool/xenstored was created every time on domain creation, that is not correct.

**Please note**, that the exact checking permissions were implemented in the last commit in order not to make one large commit and save code consistency between commits.

LKomaryanskiy commented:
As I see, CI checkpatch runs for the previous commits and prints warnings not related to PR. Also, PR has 2 expected warnings (more than 100 columns in one line), but it's because of forbidding to separate log messages.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -35,9 +37,17 @@ LOG_MODULE_REGISTER(xenstore);
+
+struct xs_permissions {
+	sys_snode_t node;
+	uint32_t domid;
```
In Xen, domid is only 16 bit wide (actually, even less), so `uint16_t` will be enough

LKomaryanskiy replied:
Fixed, changed to `domid_t`, that is `uint16_t`.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -35,9 +37,17 @@ LOG_MODULE_REGISTER(xenstore);
+struct xs_permissions {
+	sys_snode_t node;
+	uint32_t domid;
+	uint32_t perms;
```
Why not `enum xs_perm`?

LKomaryanskiy replied:
Fixed

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+		return NULL;
+	}
+
+	ptr = str;
```
You already set `ptr` to `str` at line 646

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+			goto err_free;
+		}
+
+		i += strlen(ptr + i) + 1;
```
I may miss something, but I don't see where `j` is being increased in this loop

LKomaryanskiy replied:
Oh, I missed it. Fixed

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+
+	new_perms = string_to_perms(perm_str, perms_len, &perms_num);
+	if (!new_perms) {
+		k_mutex_unlock(&xsel_mutex);
```
Do you need to unlock the mutex in this function? It looks inconsistent

LKomaryanskiy replied:
I'm sorry, it's my copy-paste issue during development. Removed.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -877,15 +1089,132 @@ static void handle_control(struct xenstore *xenstore, uint32_t id,
+		return 'w';
+	case XS_PERM_READ:
+		return 'r';
+	case (XS_PERM_READ | XS_PERM_WRITE):
```
`case XS_PERM_BOTH`?

lorc commented on xen-dom-mgmt/src/xen-dom-mgmt.c:
```diff
@@ -618,86 +618,6 @@ struct xen_domain *domid_to_domain(uint32_t domid)
-	char rbuffer[INIT_XENSTORE_BUFF_SIZE] = { 0 };
-	char uuid[40];
-	static const char basepref[] = "/local/domain";
-	static const char * const dirs[] = { "data",
```
I do not support this change. We need to separate entities. xenstore-srv should  act as a server, it should not know anything about underlying logic. 

What you will do when we'll start implementing backend/frontend configuration? There will be additional writes by each PV protocol drivers. Will you move that functionality into xenstore as well?

LKomaryanskiy replied:
Oh, I got it. I remastered corresponding patch to use a new API, add error checks and leave function as it was before. Also, I move xenstore domain deinit function to the same file.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -202,11 +202,30 @@ static int construct_path(char *payload, uint32_t domid, char **path)
  * only after all actions with entry will be performed.
  */
-static struct xs_entry *key_to_entry(const char *key)
+static struct xs_entry *key_to_entry(const char *key, uint32_t domid, uint32_t perms)
```
This is not good idea to make `key_to_entry` check permissions. Or you need to rename it to `key_to_entry_also_check_permissions` :)

LKomaryanskiy replied:
I understand, that the proposed naming was a joke, but `key_to_entry_check_perm` likely describes what function is doing. I tried to remove permissions checks from it, but it led to boilerplate code in many places:( So, I just renamed the function.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -781,6 +809,12 @@ static int xss_do_write(const char *const_path, const char *data, uint32_t domid
 
 		if (iter == NULL) {
+			if (parent_entry && !check_perms(parent_entry, XS_PERM_WRITE, domid)) {
+				LOG_ERR("Permission denied");
```
`LOG_DBG` or `LOG_INF` at most. Ot malicious guest will spam you with this messages. Also, it is nice to print which guest caused this and for which path.

LKomaryanskiy replied:
Changed loglevel and added info about path and domain.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -822,6 +856,12 @@ static int xss_do_write(const char *const_path, const char *data, uint32_t domid
 
 	if (iter && data_len > 0) {
+		if (!check_perms(iter, XS_PERM_WRITE, domid)) {
+			LOG_ERR("Permission denied");
```
The same

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -1871,6 +1871,45 @@ static void free_buffered_data(struct xenstore *xenstore)
+					prev_node = &iter_perm->node;
+				}
+			}
+			remove_domain_referenced_entries_recursive(domid, &entry->child_list);
```
nit: hope it may be shorter :)

LKomaryanskiy replied:
Renamed to shorter version.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -35,9 +37,17 @@ LOG_MODULE_REGISTER(xenstore);
  */
 #define XENSTORE_MAX_OUT_BUF_NUM 10
 
+
```
Removed

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+			goto err_free;
+		}
+
+		perms[j].domid = atoi(ptr + i + 1);
```
Unfortunately, we don't have alternative options such as `scanf`. So I decided to implement the custom function for conversion because wrong conversion to zero by `atoi` looks like a potential security issue.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+
+	free_perms(&entry->perms);
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&list_perm, perms, next, node) {
+		sys_slist_append(&entry->perms, &perms->node);
```
Added one more variable to avoid confusion.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +587,174 @@ static void free_node(struct xs_entry *entry)
+
+	*perms_num = 0;
+	ptr = str;
+	for (i = 0; i < str_len;) {
```
Added comment above the loop.

LKomaryanskiy commented:
@lorc @firscity thank you for your close review.
I released a new version of PR, so, please take a look.
All your comments are fixed.
The new PR contains the following changes compared to the previous:
1) Changed `domid` type in `xs_permissions` structure to `domid_t`;
2) Changed `perms` type in `xs_permissions` structure `enum xs_perm`;
3) Removed double initialization of iteration pointer in `string_to_perms` function;
4) Removed redundant newline;
5) Implemented helper function for safe conversion from string to `uint16_t` to avoid using `atoi` function;
6) Add missing incrementation of `j` variable in `string_to_perms`;
7) Changed some function namings to more appropriate;
8) Moved back `initialize_xenstore` function and just refactored it to use the new API ;
9) Moved `deinitialize_domain_xenstore` to `xen-dom-mgmt.c` to separate entities of xenstore client and server.

lorc commented on xenstore-srv/include/xss.h:
```diff
@@ -12,7 +12,9 @@ enum xs_perm {
-	XS_PERM_BOTH = XS_PERM_WRITE | XS_PERM_READ
+	XS_PERM_BOTH = XS_PERM_WRITE | XS_PERM_READ,
+	/* For internal implementation only */
+	XS_PERM_OWNER = 0x4,
```
It appears, that you don't need this. See my comments below about first entry semantics

LKomaryanskiy replied:
Removed

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -221,11 +221,30 @@ static int construct_path(char *payload, uint32_t domid, char **path)
+		}
+	}
+
+	return false;
```
This is not completely right, because the very first entry in permissions is describing default access rights for other domains

LKomaryanskiy replied:
Oh, I saw [this](https://wiki.xenproject.org/wiki/XenBus), but I misunderstood this. Now it's clear and it has really strange semantics. Thank you for your attention and for rechecking the exact specification. I'll rewrite the function.

lorc reviewed (changes requested):
Sorry for the new pack of the comments, but today I read the protocol specification. It is horrific :(

Especially the part about the first entry in the permissions list.

LKomaryanskiy commented:
@lorc thank you again for your close review and rechecking the specification. I apologize for misunderstanding about the first permissions entry meaning - it was really uncommon thing.

Please, take a look at the new version of PR. In comparison with the previous the following changes are made:
1) Reworked checking permissions: taking into account the value of the first permission entry, which means permissions for other domains that are not explicitly defined;
2) Changed custom function for string to integer conversion to strtoul;
3) Changed the namings of the functions to show that we are working with multiple strings separated by terminating null;
4) Added some useful comments;
5) Implemented helper function to check if domain is owner. With this function, we can omit using XS_PERM_OWNER, which was removed at all.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -877,15 +1110,132 @@ static void handle_control(struct xenstore *xenstore, uint32_t id,
+}
+
+/* The function allocates memory for the buffer, that should be freed by caller */
+static char *perms_to_string(struct xs_entry *entry, size_t *total_size)
```
Renamed

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +605,177 @@ static void free_node(struct xs_entry *entry)
+	return -ENOMEM;
+}
+
+static struct xs_permissions *string_to_perms(const char *str, const size_t str_len,
```
Renamed

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -567,15 +605,177 @@ static void free_node(struct xs_entry *entry)
+		}
+		perms->domid = new_perms[i].domid;
+		perms->perms = new_perms[i].perms;
+		/* The first permission entry is OWNER permission */
```
Now, I understand this, so yes, these lines in removed.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -153,6 +162,25 @@ static bool is_root_path(const char *path)
+ * For such conversion, we have only atoi that is
+ * not safe, so we have to use custom function.
+ */
+static int str_to_uint16(const char *str, uint16_t *num)
```
I'm sorry, I forgot about this function. Changed to strtoul.

lorc commented on xenstore-srv/include/xss.h:
```diff
@@ -12,7 +12,7 @@ enum xs_perm {
 	XS_PERM_READ = 0x1,
 	XS_PERM_WRITE = 0x2,
-	XS_PERM_BOTH = XS_PERM_WRITE | XS_PERM_READ
+	XS_PERM_BOTH = XS_PERM_WRITE | XS_PERM_READ,
```
No need in this change

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -202,11 +202,48 @@ static int construct_path(char *payload, uint32_t domid, char **path)
+{
+	struct xs_permissions *owner = SYS_SLIST_PEEK_HEAD_CONTAINER(&entry->perms, owner, node);
+
+	/* Caller is Dom0 or owner */
```
There is no check for dom0. So comment is not valid anymore.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -202,11 +202,48 @@ static int construct_path(char *payload, uint32_t domid, char **path)
  * only after all actions with entry will be performed.
  */
-static struct xs_entry *key_to_entry(const char *key)
+static struct xs_entry *key_to_entry_check_perm(const char *key, uint32_t domid, uint32_t perms)
```
I am sorry, I missed this during the previous review. It is not a good idea to add a new functionality to an already big and complex function.

I propose to leave `key_to_entry` as is (we even might need it in the future) and introduce a new function `key_to_entry_check_perm` which calls `key_to_entry` followed by `check_perms`. This will lead to a cleaner code.

lorc reviewed (commented):
With last comments addressed: 

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

LKomaryanskiy commented:
@lorc thank you for your review.
All your comments are fixed. 
 In comparison with the previous PR version the following changes are made:
1) Removed redundant comment;
2) Removed redundant comma;
3) Introduced a new function `key_to_entry_check_perm` and left `key_to_entry` as it was;
4) Added @lorc RB.

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

LKomaryanskiy commented:
@firscity thank you for your close review. Added acked-by
